qiskit_nature/operators/second_quantization/vibrational_op.py

```diff
@@ -0,0 +1,338 @@
+                        `register_length`
+            spin: positive half-integer (integer or half-odd-integer) that represents spin.
+        Raises:
+            TypeError: given data has invalid type.
+            ValueError: invalid labels.
+        """
+        if not isinstance(data, (tuple, list, str)):
+            raise TypeError(f"Type of data must be str, tuple, or list, not {type(data)}.")
+
+        if isinstance(data, tuple):
+            if not isinstance(data[0], str) or not isinstance(data[1], (int, float, complex)):
+                raise TypeError(
+                    f"Data tuple must be (str, number), not ({type(data[0])}, {type(data[1])})."
+                )
+            data = [data]
+
+        if isinstance(data, str):
+            data = [(data, 1)]
+
+        if not all(
+            isinstance(label, str) and isinstance(coeff, (int, float, complex))
+            for label, coeff in data
+        ):
+            raise TypeError("Data list must be [(str, number)].")
+
+        self._coeffs: np.ndarray
+        self._labels: List[str]
+
+        if isinstance(num_modals, int):
+            num_modals = [num_modals] * num_modes
+
+        self._num_modes = num_modes
+        self._num_modals = num_modals
+        self._register_length = sum(self._num_modals)
+
+        labels, coeffs = zip(*data)
+        self._coeffs = np.array(coeffs, np.complex128)
+
+        if not any('_' in label for label in labels):
+            # Dense label
+            if not all(len(label) == self._register_length for label in labels):
+                raise ValueError("Lengths of strings of label are different.")
+            label_pattern = re.compile(r"^[I\+\-NE]+$")
+            invalid_labels = [label for label in labels if not label_pattern.match(label)]
+            if invalid_labels:
+                raise ValueError(f"Invalid labels for dense labels are given: {invalid_labels}")
+            self._labels = list(labels)
+        else:
+            # Sparse label
+            _validate_vibrational_labels(data, num_modes, num_modals)
+            dense_labels = _convert_to_dense_labels(data, num_modes, num_modals)
+
+            ops: List[VibrationalOp] = []
+            for dense_label, coeff in dense_labels:
+                new_op = reduce(lambda a, b: a @ b, (
+                        VibrationalOp((label, 1), num_modes, num_modals) for label in dense_label
+                ))
+                # We ignore the type here because mypy only sees the complex coefficient
+                ops.append(coeff * new_op)  # type: ignore
+
+            op = reduce(lambda a, b: a + b, ops)
+
+            self._labels = op._labels.copy()
+
+    def __repr__(self) -> str:
+        if len(self) == 1:
+            if self._coeffs[0] == 1:
+                return f"VibrationalOp('{self._labels[0]}')"
+            return f"VibrationalOp({self.to_list()[0]})"
+        return f"VibrationalOp({self.to_list()})"  # TODO truncate
```

Comment:
Looking at TODO here

Reply:
Copied over from `FermionicOp`. Thus it was placed by @ikkoham.
I assume this is something we will address in the future